### DISCUSSION TITLE
Write access to a capped relationship

### DISCUSSION THREAD

**Discussion by pansen** [sqlalchemy/sqlalchemy#9052] [Usage Questions] [answered]
### Describe the use case

Hi and happy 2023 :) 

I maintain a DB schema with 1:n relationships, where the `n` part can easily become too big to load efficiently. After trying a few things I was happy to find a solution for something that came to my mind as well: [a limited relationship via Postgres window functions](https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#row-limited-relationships-with-window-functions). The description is perfect and I also got this working. 

My problem is, that I want to have this capped relation to be writable. To avoid having to load the full list when requiring write access. 

### Databases / Backends / Drivers targeted

Postgres. There is a full dev setup in the linked repo, containing a Docker Postgres and a Python project. 

### Example Use

I set up a demo project, which incorporates exactly your described solution. In [that branch](https://github.com/pansen/sqlalchemy-virtual-relationships/pull/2/files) I'm trying to write to the capped relation, which fails with 

```
E           sqlalchemy.orm.exc.UnmappedColumnError: Can't execute sync rule for destination column 'anon_1.parent_id'; mapper 'mapped class Child->child' does not map this column.  Try using an explicit `foreign_keys` collection which does not include this column (or use a viewonly=True relation).
```

That message indicates to me, that it *should* be possible. 

### Additional context

Unfortunately I could not figure out how to specify the foreign keys, to make this work. Would be awesome if you have an advice how to do so. 

Thanks! :)

**Comment by zzzeek** [comment, marked as answer]
for relationships to be writable, they generally need a simple criteria that can be translated into which columns on the target object need to be populated.

for this reason your best approach would be to make a second relationship with a simple join condition and to then use a [write only relationship](https://docs.sqlalchemy.org/en/20/orm/large_collections.html#write-only-relationships) where you can append and remove items without any load of the full collection being implied.

**Reply by pansen** [reply]
Thanks, I will work that through. 

Does that mean, there is no way to convince the capped relation to be writable?

**Reply by zzzeek** [reply]
maybe if you gave it very specific local/remote pairs, but I think it gets confusing when you get into that.    I'm not sure to the degree we've supported that pattern in the past.

**Reply by zzzeek** [reply]
that is I think the write only relationship is cleaner since you can add objects that dont correspond to the limiting condition and they will be persisted without implying any fetch access.

**Reply by pansen** [reply]
Understood. Thanks a lot!

**Reply by pansen** [reply]
Am I right, that I need to upgrade to SA 2.0? A spontaneous trial with 1.4 like this:

```python
    children_writes = relationship(
        'Child',
        order_by="desc(Child.created_at)",
        cascade="all, delete-orphan",
        passive_deletes=True,
        lazy="write_only",
    )
```

results in 
```
../../../../.venv/lib/python3.10/site-packages/sqlalchemy/orm/relationships.py:2159: in do_init
    super(RelationshipProperty, self).do_init()
../../../../.venv/lib/python3.10/site-packages/sqlalchemy/orm/interfaces.py:671: in do_init
    self.strategy = self._get_strategy(self.strategy_key)
../../../../.venv/lib/python3.10/site-packages/sqlalchemy/orm/interfaces.py:633: in _get_strategy
    cls = self._strategy_lookup(self, *key)
../../../../.venv/lib/python3.10/site-packages/sqlalchemy/orm/interfaces.py:717: in _strategy_lookup
    raise orm_exc.LoaderStrategyException(
E   sqlalchemy.orm.exc.LoaderStrategyException: Can't find strategy (('lazy', 'write_only'),) for Parent.children_writes
```

**Reply by pansen** [reply]
Ah, `lazy=noload`

```
    children_writes = relationship(
        'Child',
        order_by="desc(Child.created_at)",
        cascade="all, delete-orphan",
        passive_deletes=True,
        lazy="noload",
    )
```

https://docs.sqlalchemy.org/en/14/orm/relationship_api.html#sqlalchemy.orm.relationship.params.lazy

**Reply by CaselIT** [reply]
For `write_only` yes, you need to upgrade to v2. there is no equivalent in previous sqlalchemy versions.
The thing that's most similar is `dynamic` in previous sqlalchemy versions

**Reply by pansen** [reply]
Seems to be working with 1.4 and `noload` as described here: https://github.com/pansen/sqlalchemy-virtual-relationships/pull/2/files

**Reply by zzzeek** [reply]
noload is fine.   obviously when reading that collection, it has just whatever you put there prior to the last expiration
